doc/design/resource-dependency.md

```diff
@@ -0,0 +1,343 @@
+        "core.oam.dev/datainputs": "[{\"toFieldPaths\": [\"spec.connSecret\"],\"fromDataOutput\": \"mysql-conn-secret\"}]"
+    ```
+
+    Note that my-web won't be created since it has a DataInputs dependency to satisfy.
+
+2. Add a new CRD `DataOutput` with the following definition:
+
+    ```go
+    type DataOutput struct {
+      metav1.TypeMeta   `json:",inline"`
+      metav1.ObjectMeta `json:"metadata,omitempty"`
+
+      Spec   DataOutputSpec   `json:"spec,omitempty"`
+      Status DataOutputStatus `json:"status,omitempty"`
+    }
+
+    type DataOutputSpec struct {
+      // Resource specifies the reference of specific resource, which points to a unique resource.
+      Resource ResourceReference `json:"resource"`
+
+      // FieldPath specifies the field path of a given k8s-style resource.
+      FieldPath string `json:"fieldPath"`
+    }
+
+    type ResourceReference struct {
+      APIVersion string `json:"apiVersion"`
+
+      Kind string `json:"kind"`
+
+      // Name indicates the name of the object in the same namespace.
+      Name string `json:"name,omitempty"`
+    }
+
+    type DataOutputStatus struct {
+      // Ready indicates whether the field data of the source resource is ready.
+      Ready bool `json:"ready"`
+    }
+    ```
+
+    For example, a mysql connection secret DataOutput coming from RDSInstance my-rds would look like:
+
+    ```yaml
+    apiVersion: experimental.core.oam.dev/v1alpha2
+    kind: DataOutput
+    metadata:
+      name: mysql-conn-secret
+    spec:
+      resource:
+        apiVersion: alibaba.oam.crossplane.io/v1alpha1
+        kind: RDSInstance
+        name: my-rds
+      fieldPath: "status.connSecret"
+    ```
+
+    In this case, the OAM runtime will automatically build a dependency between the `my-web` and `my-rds`.
+
+    It will wait until `my-rds` object's field `status.connSecret` has value, and marks the DataOutput status ready:
+
+    ```yaml
+    apiVersion: alibaba.oam.crossplane.io/v1alpha1
+    kind: RDSInstance
+    metadata:
+      name: my-rds
+    spec:
+      ...
+    status:
+      connSecret: mysql-conn # OAM runtime will wait for this field to have value
+    ---
+    apiVersion: experimental.core.oam.dev/v1alpha2
+    kind: DataOutput
+    metadata:
+      name: mysql-conn-secret
+    ...
+    status:
+      ready: true # once the above connSecret has value this will be marked true
+    ```
+
+    At this point, the dependency has been satisfied.
+    The OAM runtime will create `my-web` object with corresponding values passed to its fields.
+    
+    ```yaml
+    kind: MyWorkload
+    metadata:
+      name: my-web
+      ..
+    spec:
+      connSecret: mysql-conn # patched by OAM runtime
+    ```
+
+3. Below we show how to combine the above concepts (DataInput, DataOutput) into OAM.
+
+    In OAM Component we put a syntax-sugar on existing `Parameters` fields:
+    as long as the name of the Parameter matches the name of the DataOutput, it becomes a DataInput.
```

Comment:
This seems to answer my question above, but can you also explicitly provide `dataInputs`?

Reply:
Yes. This is also put into the annotations.

Actually parameter is very Component specific. What's designed here is a generic resource dependency engine that as long as some resource has DataInputs it will build corresponding dependency.

Reply:
TODO: will update the doc

Reply:
Add a section to discuss generic engine design.